### PR TITLE
/go/store/{prolly,skip,val}: New Storage Format

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

**Review by reltuk — APPROVED** [review]
Some small comments. Looks great.

File: go/store/prolly/meta_tuple.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+func writeNewChild(ctx context.Context, ns NodeStore, level uint64, items ...nodeItem) (Node, nodePair, error) {
+	child := makeProllyNode(ns.Pool(), level, items...)
+
+	ref, err := ns.Write(ctx, child)
+	if err != nil {
+		return nil, nodePair{}, err
+	}
+
+	var meta nodePair
+	if len(items) > 0 {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/meta_tuple.go:38]: `len(items) == 0` means an empty leaf node

File: go/store/prolly/node.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	stop := nd.size() - treeLevelSize - nodeCountSize
+	start := stop - cumulativeCountSize
+	buf := nd[start:stop]
+	return val.ReadUint48(buf)
+}
+
+func (nd Node) offsets() (offs val.Offsets, itemStop val.ByteSize) {
+	stop := nd.size() - treeLevelSize - nodeCountSize - cumulativeCountSize
+	itemStop = stop - val.OffsetsSize(nd.nodeCount())
+	return val.Offsets(nd[itemStop:stop]), itemStop
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/node.go:193]: I like the second suggestion, something like 
```
type SlicedBuffer struct {
    buf  []byte
    offs []uint16
}

File: go/store/prolly/node.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+func (nd Node) leafNode() bool {
+	return nd.level() == 0
+}
+
+func (nd Node) empty() bool {
+	return len(nd) == 0 || nd.nodeCount() == 0
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/node.go:201]: we should never create a zero-length node, which also means the zero value for Node is invalid. This is a sanity check

File: go/store/prolly/node_cursor.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+}
+
+func (cur *nodeCursor) compare(other *nodeCursor) int {
+	if cur.parent != nil {
+		p := cur.parent.compare(other.parent)
+		if p != 0 {
+			return p
+		}
+	}
+	assertTrue(cur.nd.nodeCount() == other.nd.nodeCount())
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/node_cursor.go:361]: yes

File: go/store/prolly/tree_chunker.go [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+		nd, _, err := tc.createNode(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return nd, nil
+	}
+
+	// (3) This is an internal Node of the tree with a single metaPair. This is a non-canonical root, and we must walk
+	//     down until we find cases (1) or (2), above.
+	assertTrue(!tc.isLeaf() && len(tc.current) == metaPairCount)
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/tree_chunker.go:372]: we're asserting:
(1) we're at in internal node
(2) the internal node has a single metaPair (and is therefore "non-canonical")

[the remark this answers is not in the document]
Comment (reltuk) [reply go/store/prolly/tree_chunker.go]: Right, but `tc.isLeaf() == true` is impossible given the above if condition.

File: go/store/prolly/tree_chunker.go [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+			}
+		}
+
+		return tc.parent.Done(ctx)
+	}
+
+	// At this point, we know |tc.current| contains every item at this level of the tree.
+	// To see this, consider that there are two ways items can enter |tc.current|.
+	//  (1) as the result of resume() with the cursor on anything other than the first item in the Node
+	//  (2) as a result of a child treeChunker hitting an explicit chunk boundary during either Append() or finalize().
```
[the remark this answers is not in the document]
Comment (timsehn) [reply go/store/prolly/tree_chunker.go:361]: I'm surprised Aaron did not say "if the comment is wrong, delete the comment"

**Pull request merged**